From f7675b774e492074044373a3346c7dd9b9f1385b Mon Sep 17 00:00:00 2001
From: cryptite <cryptite@gmail.com>
Date: Wed, 24 Jul 2019 15:55:43 -0500
Subject: [PATCH] Read Only Worlds


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 775aca85..5484c1e4 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -2129,6 +2129,10 @@ public interface World extends PluginMessageRecipient, Metadatable {
      */
     public void setAutoSave(boolean value);
 
+    public boolean isReadOnly();
+
+    public void setReadOnly(boolean value);
+
     /**
      * Sets the Difficulty of the world.
      *
-- 
2.17.1.windows.2

